--- a/Makefile
+++ b/Makefile
@@ -4,18 +4,35 @@
 #
 PROGNAME=dump1090
 
-ifdef PREFIX
-BINDIR=$(PREFIX)/bin
-SHAREDIR=$(PREFIX)/share/$(PROGNAME)
-EXTRACFLAGS=-DHTMLPATH=\"$(SHAREDIR)\"
+BINDIR=$(DESTDIR)/usr/bin
+SBINDIR=$(DESTDIR)/usr/sbin
+SHAREDIR=$(DESTDIR)/usr/share/$(PROGNAME)
+EXTRACFLAGS=-DHTMLPATH=\"$(SHAREDIR)/public_html\"
+
+CFLAGS=-O2 -g -Wall -W $(shell pkg-config --cflags librtlsdr)
+CFLAGS += $(shell dpkg-buildflags --get CFLAGS)
+CFLAGS += $(shell dpkg-buildflags --get CPPFLAGS)
+CFLAGS += $(shell dpkg-buildflags --get CXXFLAGS)
+LIBS=$(shell pkg-config --libs librtlsdr) -lpthread -lm
+LDFLAGS=$(shell dpkg-buildflags --get LDFLAGS)
+CC=gcc
+
+ARCH=$(shell uname -m)
+
+# ppup1090 requires coaa1090.obj which is an ARM binary.
+# So only build ppup1090 on an ARM platform.
+ifneq (,$(findstring arm, $(ARCH)))
+	ARCH=arm
+else
+	ARCH=other
 endif
 
-CFLAGS=-O2 -g -Wall -W `pkg-config --cflags librtlsdr`
-LIBS=`pkg-config --libs librtlsdr` -lpthread -lm
-CC=gcc
+all: all-$(ARCH)
+install: installbin-$(ARCH) installhtml
 
 
-all: dump1090 view1090
+all-other: dump1090 view1090
+all-arm: dump1090 ppup1090 view1090
 
 %.o: %.c
 	$(CC) $(CFLAGS) $(EXTRACFLAGS) -c $<
@@ -23,8 +40,25 @@
 dump1090: dump1090.o anet.o interactive.o mode_ac.o mode_s.o net_io.o
 	$(CC) -g -o dump1090 dump1090.o anet.o interactive.o mode_ac.o mode_s.o net_io.o $(LIBS) $(LDFLAGS)
 
+ppup1090: ppup1090.o anet.o interactive.o mode_ac.o mode_s.o net_io.o
+	$(CC) -g -o ppup1090 ppup1090.o anet.o interactive.o mode_ac.o mode_s.o net_io.o coaa1090.obj $(LIBS) $(LDFLAGS)
+
 view1090: view1090.o anet.o interactive.o mode_ac.o mode_s.o net_io.o
 	$(CC) -g -o view1090 view1090.o anet.o interactive.o mode_ac.o mode_s.o net_io.o $(LIBS) $(LDFLAGS)
 
 clean:
-	rm -f *.o dump1090 view1090
+	rm -f *.o dump1090 ppup1090 view1090
+
+installbin-other:
+	install -m755 dump1090 $(SBINDIR)
+	install -m755 view1090 $(BINDIR)
+
+installbin-arm:
+	install -m755 dump1090 $(SBINDIR)
+	install -m755 ppup1090 $(SBINDIR)
+	install -m755 view1090 $(BINDIR)
+
+installhtml:
+	install -d -m 755 $(SHAREDIR)/public_html
+	cp -R public_html/* $(SHAREDIR)/public_html
+
